azkaban-common/src/test/java/azkaban/executor/DispatcherTest.java

```diff
@@ -0,0 +1,635 @@
+      this.registerFactorFilter(FactorFilter.create("requiredRemainingTmpSpace",
+          new FactorFilter.Filter<MockExecutorObject,MockFlowObject>() {
+        public boolean check(MockExecutorObject itemToCheck, MockFlowObject sourceObject) {
+          // REAL LOGIC COMES HERE -
+          if (null == itemToCheck || null == sourceObject){
+            return false;
+          }
+
+         return itemToCheck.remainingTmp > sourceObject.requiredRemainingTmpSpace;
+        }}));
+    }
+
+  }
+
+  // mock comparator class.
+  protected class MockComparator
+  extends CandidateComparator<MockExecutorObject>{
+
+    @Override
+    public String getName() {
+      return "MockComparator";
+    }
+
+    @Override
+    protected boolean differentiate(MockExecutorObject object1, MockExecutorObject object2){
+      if (object2 == null) return true;
+      if (object1 == null) return false;
+      return object1.name.compareTo(object2.name) >= 0;
+    }
+
+    public MockComparator(){
+    }
+
+    public void registerComparerForMemory(int weight){
+      this.registerFactorComparator(FactorComparator.create("Memory", weight, new Comparator<MockExecutorObject>(){
+        public int compare(MockExecutorObject o1, MockExecutorObject o2) {
+          int result = 0 ;
+
+          // check remaining amount of memory.
+          result = o1.amountOfRemainingMemory - o2.amountOfRemainingMemory;
+          if (result != 0){
+            return result > 0 ? 1 : -1;
+          }
+
+          // check remaining % .
+          result = (int)(o1.percentOfRemainingMemory - o2.percentOfRemainingMemory);
+          return result == 0 ? 0 : result > 0 ? 1 : -1;
+
+        } }));
+    }
+
+    public void registerComparerForRemainingSpace(int weight){
+      this.registerFactorComparator(FactorComparator.create("RemainingTmp", weight, new Comparator<MockExecutorObject>(){
+        public int compare(MockExecutorObject o1, MockExecutorObject o2) {
+          int result = 0 ;
+
+          // check remaining % .
+          result = (int)(o1.remainingTmp - o2.remainingTmp);
+          return result == 0 ? 0 : result > 0 ? 1 : -1;
+
+        } }));
+    }
+
+    public void registerComparerForPriority(int weight){
+      this.registerFactorComparator(FactorComparator.create("Priority", weight, new Comparator<MockExecutorObject>(){
+        public int compare(MockExecutorObject o1, MockExecutorObject o2) {
+          int result = 0 ;
+
+          // check priority, bigger the better.
+          result = (int)(o1.priority - o2.priority);
+          return result == 0 ? 0 : result > 0 ? 1 : -1;
+
+        } }));
+    }
+  }
+
+  // test samples.
+  protected ArrayList<MockExecutorObject> executorList = new ArrayList<MockExecutorObject>();
+
+  @Before
+  public void setUp() throws Exception {
+    BasicConfigurator.configure();
+
+    executorList.clear();
+    executorList.add(new MockExecutorObject("Executor1",8080,50.0,2048,5,new Date(), 20, 6400));
```

Comment:
Just thinking
Do we always need that much (12) executors to unit test our dispatching features?
Wondering if it's possible to prove certain aspects of our dispatch with less executors